nomad/encrypter.go

```diff
@@ -529,6 +505,53 @@ ERR_WAIT:
 
 }
 
+// replicateKey replicates a single key from peer servers that was present in
+// the state store but missing from the keyring. Returns an error only if no
+// peers have this key.
+func (krr *KeyringReplicator) replicateKey(ctx context.Context, keyMeta *structs.RootKeyMeta) error {
+	keyID := keyMeta.KeyID
+	krr.logger.Info("replicating new key", "id", keyID)
+
+	getReq := &structs.KeyringGetRootKeyRequest{
+		KeyID: keyID,
+		QueryOptions: structs.QueryOptions{
+			Region:        krr.srv.config.Region,
+			MinQueryIndex: keyMeta.ModifyIndex - 1,
+		},
+	}
+	getResp := &structs.KeyringGetRootKeyResponse{}
+	err := krr.srv.RPC("Keyring.Get", getReq, getResp)
+
+	if err != nil || getResp.Key == nil {
+		// Key replication needs to tolerate leadership flapping. If a key is
+		// rotated during a leadership transition, it's possible that the new
+		// leader has not yet replicated the key from the old leader before the
+		// transition. Ask all the other servers if they have it.
+		krr.logger.Warn("failed to fetch key from current leader, trying peers",
+			"key", keyID, "error", err)
+		getReq.AllowStale = true
+		for _, peer := range krr.getAllPeers() {
+			err = krr.srv.forwardServer(peer, "Keyring.Get", getReq, getResp)
+			if err == nil && getResp.Key != nil {
+				break
+			}
+		}
+		if getResp.Key == nil {
+			krr.logger.Error("failed to fetch key from any peer",
+				"key", keyID, "error", err)
+			return fmt.Errorf("failed to fetch key from any peer: %v", err)
+		}
+	}
+
+	err = krr.encrypter.AddKey(getResp.Key)
+	if err != nil {
+		return fmt.Errorf("failed to add key to keyring: %v", err)
+	}
+
+	krr.logger.Info("added key", "key", keyID)
```

Comment:
Do we need before and after log lines at Info level? If knowing how long replication takes is critical it seems reasonable, but I'd have trouble explaining to folks why they need to see these.

Reply:
Eh, probably not. I'll be honest and say I got a little overzealous with the logging here because the internal customer was not collecting any debug-level logs. But they're also not seeing _any_ logs from this code path for some reason (something to do with their collection, I think), so that really is just excessive. Will fix.

Reply:
Done